whelktool/src/main/groovy/whelk/datatool/form/MatchForm.groovy

```diff
@@ -0,0 +1,307 @@
+package whelk.datatool.form
+
+import groovy.transform.Memoized
+import whelk.Document
+import whelk.JsonLd
+import whelk.Whelk
+import whelk.datatool.util.DocumentComparator
+import whelk.datatool.util.IdLoader
+import whelk.util.DocumentUtil
+
+import static whelk.JsonLd.GRAPH_KEY
+import static whelk.JsonLd.ID_KEY
+import static whelk.JsonLd.RECORD_KEY
+import static whelk.JsonLd.RECORD_TYPE
+import static whelk.JsonLd.THING_KEY
+import static whelk.JsonLd.TYPE_KEY
+import static whelk.JsonLd.asList
+import static whelk.component.SparqlQueryClient.GRAPH_VAR
+import static whelk.converter.JsonLDTurtleConverter.toTurtleData
+import static whelk.util.DocumentUtil.getAtPath
+import static whelk.util.LegacyIntegrationTools.getMarcCollectionInHierarchy
+
+class MatchForm {
+    private static final DocumentComparator comparator = new DocumentComparator()
+
+    public static final String MATCHING_MODE = 'bulk:matchingMode'
+    public static final String HAS_ID = 'bulk:hasId'
+    public static final String BNODE_ID = 'bulk:formBlankNodeId'
+    public static final String ANY_TYPE = "bulk:Any"
+    public static final String SUBTYPES = "bulk:Subtypes"
+    public static final String EXACT = 'bulk:Exact'
+
+    private static final String VALUE = 'value'
+    private static final String VALUE_FROM = 'bulk:valueFrom'
+    private static final String ANY_OF = 'bulk:AnyOf'
+    private static final String HAS_BASE_TYPE_TMP = '_hasBaseTypeTmp'
+
+    Map form
+
+    Map<String, List> nodeIdToMatchFormPath
+    Map<String, Set<String>> nodeIdMappings
+    Map<String, Set<String>> baseTypeMappings
+
+    MatchForm(Map form, Whelk whelk) {
+        this.form = form
+        this.nodeIdToMatchFormPath = collectNodeIdToPath(form)
+        this.nodeIdMappings = collectNodeIdMappings(whelk)
+        this.baseTypeMappings = collectBaseTypeMappings(whelk?.jsonld)
+    }
+
+    MatchForm(Map matchForm) {
+        this(matchForm, null)
+    }
+
+    MatchForm() {}
+
+    Map<String, Set<String>> collectNodeIdMappings(Whelk whelk) {
+        return collectNodeIdMappings(form, whelk)
+    }
+
+    static Map<String, Set<String>> collectNodeIdMappings(Map form, Whelk whelk) {
```

Comment:
I think a short comment would be helpful here.
Since the method name is not self explanatory.
What is the result? `Map<String, Set<String>>`

Reply:
Or rather: descibe the member variables
```
    Map<String, List> nodeIdToMatchFormPath
    Map<String, Set<String>> nodeIdMappings
    Map<String, Set<String>> baseTypeMappings
```